include/deal.II/multigrid/mg_transfer_interface.h

```diff
@@ -0,0 +1,424 @@
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_mg_transfer_interface_h
+#define dealii_mg_transfer_interface_h
+
+#include <deal.II/base/mg_level_object.h>
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/lac/affine_constraints.h>
+#include <deal.II/lac/la_parallel_vector.h>
+
+#include <deal.II/multigrid/mg_base.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declarations
+#ifndef DOXYGEN
+namespace MGTransferUtil
+{
+  class Implementation;
+}
+#endif
+
+/**
+ * A multigrid transfer scheme. A multrigrid transfer class can have different
+ * transfer schemes to enable p-adaptivity (one transfer scheme per
+ * polynomial degree pair) and to enable global coarsening (one transfer
+ * scheme for transfer between children and parent cells, as well as, one
+ * transfer scheme for cells that are not refined).
+ *
+ * @note Normally, this data structure does not to be filled by users, but, one
+ *   can use the can use the utility functions provided in the namespace
+ *   MGTransferUtil to setup the Transfer operators.
+ */
+template <typename Number>
+struct TransferScheme
+{
+  /**
+   * Number of coarse cells.
+   */
+  unsigned int n_cells_coarse;
+
+  /**
+   * Number of degrees of freedom of a coarse cell.
+   */
+  unsigned int n_cell_dofs_coarse;
+
+  /**
+   * Number of degrees of freedom of fine cell.
+   */
+  unsigned int n_cell_dofs_fine;
+
+  /**
+   * Polynomial degree of the finite element of the coarse cells.
+   */
+  unsigned int degree_coarse;
+
+  /**
+   * Polynomial degree of the finite element of the fine cells.
+   */
+  unsigned int degree_fine;
+
+  /**
+   * Flag if the finite element on the fine cells are contiguous. If yes,
+   * the multiplicity of DoF shares and constraints have to be taken in account
+   * via weights.
+   */
+  bool fine_element_is_continuous;
+
+  /**
+   * Weights for continuous elements.
+   */
+  std::vector<Number> weights;
+
+  /**
+   * 1D prolongation matrix.
+   */
+  AlignedVector<VectorizedArray<Number>> prolongation_matrix_1d;
+
+  /**
+   * DoF indices of the coarse cells.
+   */
+  std::vector<unsigned int> level_dof_indices_coarse;
+
+  /**
+   * DoF indices of the children of the coarse cells.
+   */
+  std::vector<unsigned int> level_dof_indices_fine;
+
+  /**
+   * Print internal data structures to stream @p out.
+   */
+  template <typename Stream>
+  void
+  print(Stream &out) const;
+};
+
+
+
+/**
+ * Class for transfer between two multigrid levels.
+ */
+template <int dim, typename Number>
+class Transfer
```

Comment:
Most of our multigrid related classes start with ``MG...``. How does this class relate to ``MGTransfer*``?

Reply:
What about `MGTwoLevelTransfer`? It is the transfer operator between two multigrid leveles (however they are defined).